### PR TITLE
Add scaleway provider

### PR DESCRIPTION
**Pull request by nicolai86**
This PR adds a provider for the [scaleway ARM cloud hosting provider](https://www.scaleway.com/).

One can manage the entire scaleway stack like this:

```
provider "scaleway" {
  api_key = "snap"
  organization = "snip"
}

resource "scaleway_ip" "base" {
  server = "${scaleway_server.base.id}"
}

resource "scaleway_server" "base" {
  name = "test"
  # ubuntu 14.04
  image = "aecaed73-51a5-4439-a127-6d8229847145"
  type = "C2S"
}

resource "scaleway_volume" "test" {
  name = "test"
  size_in_gb = 20
  type = "l_ssd"
}

resource "scaleway_volume_attachment" "test" {
  server = "${scaleway_server.base.id}"
  volume = "${scaleway_volume.test.id}"
}

resource "scaleway_security_group" "base" {
  name = "public"
  description = "public gateway"
}

resource "scaleway_security_group_rule" "http-ingress" {
  security_group = "${scaleway_security_group.base.id}"

  action = "accept"
  direction = "inbound"
  ip_range = "0.0.0.0/0"
  protocol = "TCP"
  port = 80
}


resource "scaleway_security_group_rule" "http-egress" {
  security_group = "${scaleway_security_group.base.id}"

  action = "accept"
  direction = "outbound"
  ip_range = "0.0.0.0/0"
  protocol = "TCP"
  port = 80
}
```

The PR includes tests & documentation. 

Looking forward to your feedback!

### CODE REVIEW

**Comment by nicolai86** [comment]
Volume attachments are now working properly, too.

**Comment by nicolai86** [comment]
Test output (cleaned up): 
```

TF_ACC=1 go test ./builtin/providers/scaleway -v -run=TestAccScaleway -timeout 120m
=== RUN   TestAccScalewayIP_Basic
--- PASS: TestAccScalewayIP_Basic (1.96s)
=== RUN   TestAccScalewaySecurityGroupRule_Basic
--- PASS: TestAccScalewaySecurityGroupRule_Basic (2.77s)
=== RUN   TestAccScalewaySecurityGroup_Basic
--- PASS: TestAccScalewaySecurityGroup_Basic (1.94s)
=== RUN   TestAccScalewayServer_Basic
--- PASS: TestAccScalewayServer_Basic (123.19s)
=== RUN   TestAccScalewayVolumeAttachment_Basic
--- PASS: TestAccScalewayVolumeAttachment_Basic (325.58s)
=== RUN   TestAccScalewayVolume_Basic
--- PASS: TestAccScalewayVolume_Basic (2.61s)
PASS
ok    github.com/hashicorp/terraform/builtin/providers/scaleway  458.068s
```

File: builtin/providers/scaleway/helpers.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+			return err
+		}
+		if currentState != server.State {
+			fmt.Printf("Server changed state to %q\n", server.State)
+			currentState = server.State
+		}
+		if server.State == targetState {
+			break
+		}
+		time.Sleep(1 * time.Second)
```
Comment (stack72) [inline builtin/providers/scaleway/helpers.go]: is there any issue with 1 second refresh times? I.e. will we hit their API limit or anything?

File: builtin/providers/scaleway/resource_scaleway_ip.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+	d.Set("ip", resp.IP.Address)
+	d.Set("server", resp.IP.Server.Identifier)
+	return nil
+}
+
+func resourceScalewayIPUpdate(d *schema.ResourceData, m interface{}) error {
+	scaleway := m.(*Client).scaleway
+	if d.HasChange("server") {
+		if err := scaleway.AttachIP(d.Id(), d.Get("server").(string)); err != nil {
```
Comment (stack72) [inline builtin/providers/scaleway/resource_scaleway_ip.go]: As server is `Optional` here, what happens if it is nil?

File: builtin/providers/scaleway/resource_scaleway_ip.go [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	}
+
+	d.SetId(resp.IP.ID)
+	return resourceScalewayIPUpdate(d, m)
+}
+
+func resourceScalewayIPRead(d *schema.ResourceData, m interface{}) error {
+	scaleway := m.(*Client).scaleway
+	resp, err := scaleway.GetIP(d.Id())
+	if err != nil {
```
Comment (stack72) [inline builtin/providers/scaleway/resource_scaleway_ip.go]: Can we handle the issue when someone deletes an IP from the console / cli? We should do something like

```
if err != nil {
    if resp.StatusCode == 404 {
        d.SetId("")
        log.Printf("[DEBUG] IP Not Found so refreshing from state")
        return nil
    }
}
```

File: builtin/providers/scaleway/resource_scaleway_security_group.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		Description:  d.Get("description").(string),
+	}
+
+	if err := scaleway.PutSecurityGroup(req, d.Id()); err != nil {
+		log.Printf("[DEBUG] Error reading security group: %q\n", err)
+
+		return err
+	}
+
+	return nil
```
Comment (stack72) [inline builtin/providers/scaleway/resource_scaleway_security_group.go]: I *think* we should call the Read func here - that way we can make sure that we set anything that changes back to state

File: builtin/providers/scaleway/resource_scaleway_security_group_rule.go [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+		DestPortFrom: d.Get("port").(int),
+	}
+
+	if err := scaleway.PutSecurityGroupRule(req, d.Get("security_group").(string), d.Id()); err != nil {
+		log.Printf("[DEBUG] error updating Security Group Rule: %q", err)
+
+		return err
+	}
+
+	return nil
```
Comment (stack72) [inline builtin/providers/scaleway/resource_scaleway_security_group_rule.go]: Think we should call the Read func here

File: builtin/providers/scaleway/resource_scaleway_server.go [hunk 6]
```diff
@@ -0,0 +1,183 @@
...
+
+	if d.HasChange("dynamic_ip_required") {
+		req.DynamicIPRequired = Bool(d.Get("dynamic_ip_required").(bool))
+	}
+
+	if err := scaleway.PatchServer(d.Id(), req); err != nil {
+		return fmt.Errorf("Failed patching scaleway server: %q", err)
+	}
+
+	return nil
```
Comment (stack72) [inline builtin/providers/scaleway/resource_scaleway_server.go]: Again call the Read func here

**Comment by stack72** [comment]
Hi @nicolai86 

I left a small comments - nothing major. If we can get them fixed up then that will be fantastic.

I need to try and get in touch with Scaleway to see if we can get a testing account for terraform

Paul

**Comment by nicolai86** [comment]
Hey @stack72 

thank you for your feedback. Will adjust this soonish.

Best,
Raphael

File: builtin/providers/scaleway/helpers.go (hunk 1 above)
Comment (nicolai86) [inline builtin/providers/scaleway/helpers.go]: as far as I know there's no api limit. But maybe @moul can give some insight here.

File: builtin/providers/scaleway/resource_scaleway_ip.go (hunk 2 above)
Comment (nicolai86) [inline builtin/providers/scaleway/resource_scaleway_ip.go]: It should detach the server. I'll add an explicit test case to make sure this behaves as expected. Sadly the scaleway cli does not expose a top level function like `detachIP`

File: builtin/providers/scaleway/resource_scaleway_ip.go (hunk 3 above)
Comment (nicolai86) [inline builtin/providers/scaleway/resource_scaleway_ip.go]: You can do this like this:

```
if serr, ok := err.(api.ScalewayAPIError); ok {
  if serr.StatusCode == 404 {
    d.SetId("")
    log.Printf("[DEBUG] IP Not Found when refreshing state")
    return nil
  }
}
```

Will be in the next commit.

**Comment by nicolai86** [comment]
@stack72 I think I adjusted everything based on your feedback. Ready for another review I'd say.

**Comment by stack72** [comment]
Hi @nicolai86

This looks good to me now. I just need to get an account on scale way to test this out and we are GTG

Paul

File: website/source/docs/providers/scaleway/index.html.markdown [hunk 7]
```diff
@@ -0,0 +1,96 @@
+---
+layout: "scaleway"
+page_title: "Provider: Scaleway"
+sidebar_current: "docs-scaleway-index"
+description: |-
+  The Docker provider is used to interact with Docker containers and images.
```
Comment (stack72) [inline website/source/docs/providers/scaleway/index.html.markdown]: This description should be for Scaleway :)

File: website/source/docs/providers/scaleway/index.html.markdown [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+---
+
+# Scaleway Provider
+
+The Scaleway provider is used to manage Scaleway resources.
+
+Use the navigation to the left to read about the available resources.
+
+<div class="alert alert-block alert-info">
```
Comment (stack72) [inline website/source/docs/providers/scaleway/index.html.markdown]: We can drop this block :)

File: website/source/docs/providers/scaleway/index.html.markdown [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+  action = "accept"
+  direction = "inbound"
+  ip_range = "0.0.0.0/0"
+  protocol = "TCP"
+  dest_port_from = 443
+}
+
+```
+
+You'll need to provide your Scaleway organization and Auth token,
```
Comment (stack72) [inline website/source/docs/providers/scaleway/index.html.markdown]: So I got confused here on what an organisation key was. It took me a while to realise that it was actually `access_key` - I think we should rename the env var to that.

**Comment by stack72** [comment]
Hi @nicolai86 

left 2 more small points on the docs side of things. Just managed to test this and the tests work as expected :)

The tests are very chatty though:

```
=== RUN   TestAccScalewaySecurityGroup_Basic
2016/07/13 10:46:30 POST /security_groups
2016/07/13 10:46:30 GET /security_groups
2016/07/13 10:46:30 GET /security_groups/4d25daae-04f8-4dd6-b778-e3d234dafaf0
2016/07/13 10:46:31 GET /security_groups/4d25daae-04f8-4dd6-b778-e3d234dafaf0
2016/07/13 10:46:31 GET /security_groups/4d25daae-04f8-4dd6-b778-e3d234dafaf0
2016/07/13 10:46:31 GET /security_groups/4d25daae-04f8-4dd6-b778-e3d234dafaf0
2016/07/13 10:46:31 GET /security_groups/4d25daae-04f8-4dd6-b778-e3d234dafaf0
2016/07/13 10:46:31 DELETE /security_groups/4d25daae-04f8-4dd6-b778-e3d234dafaf0
--- PASS: TestAccScalewaySecurityGroup_Basic (1.56s)
```

Is there any way that this can be disabled? I.e can they be moved within our logging levels of INFO, DEBUG, TRACE and ERROR etc?

P.

**Comment by nicolai86** [comment]
@stack72 I've updated the docs, renamed the `api_key` to cleary communicate the usage, and also made sure the logs are less chatty :)

**Comment by stack72** [comment]
Hi @nicolai86 

This now looks great and the tests pass as expected. Thanks for making those logging changes!

Paul

**Comment by nicolai86** [comment]
Hey @stack72 

thanks! If there are any bugs/ problems related to this feel free to pull me into the discussion.

Raphael

**Comment by stack72** [comment]
Will do! Thanks for all the help here :)

**Comment by sheerun** [comment]
@nicolai86 For some reason remote-exec provisioner is not working with scaleway resouces. It's stuck in endless loop:

```
scaleway_server.tunnel: Still creating... (3m20s elapsed)
scaleway_server.tunnel (remote-exec): Connecting to remote host via SSH...
scaleway_server.tunnel (remote-exec):   Host:
scaleway_server.tunnel (remote-exec):   User: root
scaleway_server.tunnel (remote-exec):   Password: false
scaleway_server.tunnel (remote-exec):   Private key: false
scaleway_server.tunnel (remote-exec):   SSH Agent: true
```

**Comment by nicolai86** [comment]
@sheerun I've got a more [complex example of the scaleway provider](https://github.com/nicolai86/scaleway-terraform-demo) up on github which works just fine with remote provisioner.

Assuming you can SSH into your instance this should work just fine

**Comment by nicolai86** [comment]
To summarize the use of `remote-exec`: 

**option A**  
use of `dynamic_ip_required`

```
resource "scaleway_server" "server" {
  image               = "${var.image}"
  type                = "${var.type}"
  dynamic_ip_required = true

  provisioner "remote-exec" {
    inline = "echo hello world"
  }
}
```

this works because setting `dynamic_ip_required = true` gives us a public IP to connect to.

**option B**  
assuming you have a jump host with a public accessible IP (see option A to create one)
use the jump host to connect:

```
resource "scaleway_server" "server" {
  image = "${var.image}"
  type  = "${var.type}"

  connection {
    type         = "ssh"
    user         = "root"
    host         = "${self.private_ip}"
    bastion_host = "${var.bastion_host}"
    bastion_user = "root"
    agent        = true
  }

  provisioner "remote-exec" {
    inline = "echo hello world"
  }
}
```

**Pull request merged**